src/main/java/com/snowflake/kafka/connector/internal/streaming/StreamingClientProvider.java

```diff
@@ -0,0 +1,236 @@
+ * provide multiple clients
+ */
+public class StreamingClientProvider {
+  private static class StreamingClientProviderSingleton {
+    private static final StreamingClientProvider streamingClientProvider =
+        new StreamingClientProvider();
+  }
+
+  /**
+   * Gets the current streaming provider
+   *
+   * @return The streaming client provider
+   */
+  public static StreamingClientProvider getStreamingClientProviderInstance() {
+    return StreamingClientProviderSingleton.streamingClientProvider;
+  }
+
+  /**
+   * Checks if the client is valid by doing a null check, ensuring it is open and has a name
+   *
+   * @param client The client to validate
+   * @return If the client is not null, open and has a name
+   */
+  public static boolean isClientValid(SnowflakeStreamingIngestClient client) {
+    return client != null && !client.isClosed() && client.getName() != null;
+  }
+
+  /**
+   * ONLY FOR TESTING - to get a provider with injected properties
+   *
+   * @param createdClientId the number of times a client has been created
+   * @param connectorConfig the connector config
+   * @param client the injected streaming client
+   * @return a provider with the injected properties
+   */
+  @VisibleForTesting
+  public static StreamingClientProvider injectStreamingClientProviderForTests(
+      int createdClientId,
+      Map<String, String> connectorConfig,
+      SnowflakeStreamingIngestClient client) {
+    return new StreamingClientProvider(createdClientId, connectorConfig, client);
+  }
+
+  private static final KCLogger LOGGER = new KCLogger(StreamingClientProvider.class.getName());
+  private static final String STREAMING_CLIENT_PREFIX_NAME = "KC_CLIENT_";
+  private final Lock clientLock = new ReentrantLock(true);
+
+  private AtomicInteger createdClientId;
+  private Map<String, String> connectorConfig;
+  private SnowflakeStreamingIngestClient streamingIngestClient;
+
+  // private constructor for singleton
+  private StreamingClientProvider() {
+    this.createdClientId = new AtomicInteger(0);
+    this.connectorConfig = new HashMap<>();
+  }
+
+  // ONLY FOR TESTING - private constructor to inject properties for testing
+  @VisibleForTesting
+  private StreamingClientProvider(
+      int createdClientId,
+      Map<String, String> connectorConfig,
+      SnowflakeStreamingIngestClient client) {
+    this.createdClientId = new AtomicInteger(createdClientId);
+    this.connectorConfig = connectorConfig;
+    this.streamingIngestClient = client;
+  }
+
+  /**
+   * Creates a new streaming client, will replace the existing one if necessary
+   *
+   * @param connectorConfig The connector config to define the client
+   */
+  public void createOrReplaceClient(Map<String, String> connectorConfig) {
+    // replace previous connector config and client if applicable
+    this.clientLock.lock();
+    LOGGER.info("Creating new client, this will replace the old client and config if exists");
```

Comment:
this can be confusing, we are loggin creating new client and below in closeClient we are printing `Closing...`
Do we need a log line here?

Reply:
removed force replace logic